src/Compilers/Core/Portable/Emit/EditAndContinue/DeltaMetadataWriter.cs

```diff
@@ -784,6 +810,192 @@ private void PopulateEncLogTableParameters()
+            }
+
+            // The data in _previousGeneration.CustomAttributesAdded is not nicely sorted, or even necessarily contiguous
+            // so we need to map each target onto the rows its attributes occupy so we know which rows to update
+            var attributeMap = CreateExistingAttributeMap(_previousGeneration.OriginalMetadata.MetadataReader, _previousGeneration.CustomAttributesAdded, out var lastRowId);
+
+            int rowId;
+            var nextRowId = lastRowId + 1;
+            var logIndex = 0;
+            foreach (var customAttributeTarget in remainingEmittedAttributeParents)
+            {
+                if (attributeMap.TryGetValue(customAttributeTarget, out var queue) &&
+                    queue.Count > 0)
+                {
+                    // If there is already an attribute for this parent, then pull the next available row number
+                    // off the queue and update it
+                    rowId = queue.Dequeue();
+                }
+                else
+                {
+                    // Otherwise, either this is a new parent that hasn't had an attribute before, or we've run
+                    // out of existing rows to update. Either way, we want to add to the end of the table.
+                    rowId = nextRowId++;
+
+                    // Keep track of the addition for passing to the next generation
+                    _customAttributesAdded[rowId] = customAttributeTarget;
+                }
+
+                // remainingEmittedAttributeParents only contains records that haven't been logged, but since we know it is in the
+                // right order, we don't need to know what the original index we can just fill log rows in order, but and skip any
+                // that were already filled in the previous phase.
+                Debug.Assert(logIndex < encLogRows.Length, "encLogRows cannot be larger than remainingEmittedAttributeParents.");
+                while (encLogRows[logIndex] != 0)
+                {
+                    logIndex++;
+                }
+
+                encLogRows[logIndex++] = rowId;
+            }
+        }
+
+        /// <summary>
+        /// For each custom attribute target that was emitted in previous generations, we need to keep a list of
+        /// which row numbers are used because whilst each generation is sorted by parent, we're looking across multiple
+        /// generations.
+        /// </summary>
+        private static Dictionary<EntityHandle, Queue<int>> CreateExistingAttributeMap(MetadataReader metadataReader, IReadOnlyDictionary<int, EntityHandle> customAttributesAdded, out int lastRowId)
+        {
+            var result = new Dictionary<EntityHandle, Queue<int>>(HandleComparer.Default);
+
+            lastRowId = metadataReader.GetTableRowCount(TableIndex.CustomAttribute);
+
+            foreach (var pair in customAttributesAdded)
+            {
+                lastRowId = Math.Max(lastRowId, pair.Key);
```

Comment:
>pair.Key [](http://example.com/codeflow?start=48&length=8)

Consider creating locals for these: `int index = pair.Key; EntityHandle parent = pair.Value;`